[클린코드 7기 이상혁] 자동차 경주 미션 STEP 5

sanghyuk-2i opened the pull request:
안녕하세요, 유조님!

5단계 미션을 최대한 많이 수행했으며 리뷰 요청 드립니다!

일단, 요구사항 이외에도 명칭이나 중복 코드를 중심으로 최대한 줄이려고 노력해 봤습니다.

## 요구사항

- [x] 도메인 로직은 domain/ 하위로, UI 관련 로직은 view/ 하위에서 관리한다.
- [x] domain/ 하위의 모듈은 view/ 하위의 모듈을 의존하지 않아야 한다.
- [x] 도메인 로직 내에서 테스트하기 어려운 부분을 분리하고, 테스트 가능한 부분에 대해서만 단위 테스트를 작성한다.
       (테스트 코드에서 jest.fn()을 사용하지 않는다.)

## 질문사항

1. 세 번째 요구 사항을 보면 "jest.fn()을 사용하지 않는다"라는 의미가 모킹을 이용하지 않고 최대한 실제 함수를 사용해서 테스트 효과를 높이라는 뜻인가요? 만약, random 로직 같은 테스트는 어쩔 수 없이 모킹으로 테스트해야 한다는 이야기는 들어서 이 부분은 약간 흐지부지 넘어가게 되었습니다.

sanghyuk-2i committed:
refactor: 상수 정리

- class static 속성으로 상수화
- 테스트 코드에 불필요해진 상수값 제거

sanghyuk-2i committed:
refactor: UI 로직 분리 및 개선

- Game 관련 UI 로직을 views 폴더로 이동
- announce에서 render로 UI 함수명 변경

sanghyuk-2i committed:
refactor: 테스트 코드 리팩토링

- 반복적인 명세서 형태를 test.each()로 리팩토링

sanghyuk-2i committed:
refactor: main 관련 UI 로직 분리

- Input(readline)도 UI라고 판단되며 UI 로직 분리

yujo11 commented on __tests__/Car.test.js:
```diff
@@ -17,12 +16,12 @@ describe('자동차 클래스 테스트', () => {
     const location = car.getLocation();
 
-    expect(location).toBe(DEFAULT_CAR_LOCATION);
+    expect(location).toBe(Car.DEFAULT_CAR_LOCATION);
```
상수를 재활용 할 수 있는 형태로 변경하셨군요 👍 👍 👍

yujo11 commented on __tests__/Game.test.js:
```diff
@@ -137,21 +135,16 @@ describe('게임(자동차 경주) 진행에 대한 테스트', () => {
-
-    expect(track).toBe('-----');
-  });
+  test.each([
+    [Car.DEFAULT_CAR_LOCATION, ''],
+    [Car.DEFAULT_CAR_LOCATION + 1, '-'],
+    [Car.DEFAULT_CAR_LOCATION + 5, '-----'],
+  ])(
```
오 `.each()` 메서드를 활용해주셨군요 아주 좋네용 ㅎㅎ 👍 👍 👍

yujo11 commented on __tests__/utils/isLapValid.test.js:
```diff
@@ -1,16 +1,13 @@
+    ['7', true],
+    ['십일', false],
+  ])('"%s"을(를) 입력하면 %s을(를) 반환한다.', (input, expected) => {
+    const isValid = isLapVaild(input);
```
소소하게 함수 명에 오타가 있네용 ㅎㅎ
```suggestion
    const isValid = isLapValid(input);
```

vsc 사용하신다면 오타 잡아주는 [이런 익스텐션](https://marketplace.cursorapi.com/items?itemName=streetsidesoftware.code-spell-checker)도 추천 드려봅니다 ㅎㅎ

yujo11 replied:
`isLapVaild` 함수의 경우 들어온 input이 숫자로 변환 가능한지 검사해주는 함수로 보이는데요.

성공 케이스와 실패케이스를 명시하고 나눠서 테스트하면 어떤 역할을 하는 함수이고, 어떤 상황에서 어떤 값을 반환하는지 더 명확히 보여서 좋을거 같네용

ex)

- number로 변환 가능한 string이 들어온 경우 true를 반환한다.
- number로 변환 불가능한 string이 들어온 경우 false를 반환한다.

sanghyuk-2i replied:
명세 내용을 수정하면서 자연스럽게 예상치 못한 경우의 수도 발견해서 
그 점을 테스트 코드와 함께 반영해 봤습니다!

- 사용자가 숫자를 입력해도 0, -1, 1.2를 입력할 수 있는 케이스
- 양의 정수 or 공백이면 true, 그게 아니면 false로 반환하도록 변경

yujo11 replied:
오오 추가적인 개선까지 진행해주셨군요 !! 👍 👍 👍

yujo11 commented on src/utils/index.js:
```diff
@@ -1,10 +1,16 @@
+const DEFAULT_MIN_NUMBER = 0;
+const DEFAULT_MAX_NUMBER = 9;
```
상수를 사용하는데는 크게 2가지 목적이 있다고 생각하는데요.

1. 의미없는 값에 이름를 부여해서 의미를 나타내게 함
2. 같은 값이 쓰이는 여러 곳에서 사용해서 유지보수를 용이하게 함

이 경우에는 `DEFAULT_`가 왜 Default인지 알기 어렵기 때문에 여전히 유의미한 의미를 나타내지 못 하고 있고, 한 군데에서만 쓰여 유지보수의 편의성도 가져다주지 못 할거 같아요.

magic number  사용보다는 좋지만 이런 부분도 같이 고려해보시면 더 좋을거 같네용 ! 상혁님은 어떤걸 기준으로 상수화를 하고 계신지도 궁금합니다 !

sanghyuk-2i replied:
말씀해 주신 내용과 어느 정도 겹치는 부분들이 있는데,

저는 
- 공통적으로 쓰이는 상황이 빈번하면
- 분기 처리하는데 왜 이 값인지 설명이 필요하면

상수를 사용하고 있습니다. 아마 Step 5가 리팩토링이다 보니까, 뭐라도 수정이 필요할 것을 찾다가 너무 일반적인 숫자들을 상수하니까 코드만 복잡해진 것 같습니다 😅

yujo11 commented on src/view/game/index.js:
오 view에 해당하는 함수들을 나눠주셨군요 좋네용 ㅎㅎ 👍 👍 👍

yujo11 reviewed (changes requested):
안녕하세요 상혁님~~ 

바로 merge 해도 될만한 코드라고 생각하지만 같이 생각해보면 좋을거 같은 부분들이 있어 소소하게 코멘트 남겼습니다. 확인해주시면 좋을거 같아요 !

>세 번째 요구 사항을 보면 "jest.fn()을 사용하지 않는다"라는 의미가 모킹을 이용하지 않고 최대한 실제 함수를 사용해서 테스트 효과를 높이라는 뜻인가요? 만약, random 로직 같은 테스트는 어쩔 수 없이 모킹으로 테스트해야 한다는 이야기는 들어서 이 부분은 약간 흐지부지 넘어가게 되었습니다.

제가 요구사항을 만든건 아니라서 정확히는 모르겠으나 그런 의도가 담겨있으리라 봅니다 ㅎㅎ 

추가로 테스트가 가능한 함수나 메서드 구조에 대해서 고민을 해보라는 의도도 있지 않나 싶은데요(물론 저의 뇌피셜입니다)

상혁님이 흐지부지 넘어갔다고 말씀해주신 코드는 아마도 랜덤 넘버를 생성해서 사용하고 있는 아래의 함수일거 같아요.

```js
  determineCarMovement(car, onMove, onStay) {
    const randomNumber = getRandomNumber();
    if (randomNumber < Game.MIN_MOVEMENT_THRESHOLD) {
      return onStay(car);
    }

    return onMove(car);
  }
```

현재는 랜덤한 숫자를 `determineCarMovement` 내부에서 생성하고 있고, 그에 따라 내부 함수의 동작이 달라져 mocking 없이는 테스트가 곤란할 수 있는 상황으로 보이는데요.

`determineCarMovement`를 mocking 없이 테스트 가능하게 만드려면 어떻게 해볼 수 있을까요? 이 부분도 같이 고민해보면 좋을거 같습니다 ㅎㅎ

---

코멘트 확인하고 다시 리뷰 요청 부탁드립니다!! 고생 많으셨습니다 ~~!! 👍 👍 👍

sanghyuk-2i committed:
fix: isLapValid 함수 수정

- 네이밍 오타 수정
- 단순히 숫자가 아닌 양의 정수인지 판별하면 boolean 값 반환하도록 처리
- 테스트 코드 context 추가 및 보완

sanghyuk-2i committed:
refactor: determineCarMovement 함수 수정

- 테스트 용이(mocking 없이)를 위한 리팩토링
- 매개변수를 통해서 임계치를 지정할 수 있도록 변경

sanghyuk-2i commented:
말씀해 주신대로, mocking 없이 해당 함수를 실행할 수 있는 구조로 어떻게 변경할지 고민한 결과...
움직임을 결정해 주는 임계치를 매개 변수로 통해서 주면 좋을 것 같다고 생각했습니다.

그래서, 테스트 코드에서는 원하는 값을 넣으면서 함수를 실행 시킬 수 있고, 
프로덕트 코드에서는 랜덤 값을 넣으면서 요구사항을 지킬 수 있었습니다.

코드 컨벤션에 따라 매개 변수를 최대 2개까지 처리하도록 작성되어 있는데, 
Callback 함수들을 구조 분해 할당으로 전달하기는 뭔가 제 스타일(?)이 아니라서 😅 일단 아래처럼 변경했습니다!

```js
 determineCarMovement(
    { car, threshold = Game.MIN_MOVEMENT_THRESHOLD },
    onMove,
    onStay,
  ) {
    if (threshold < Game.MIN_MOVEMENT_THRESHOLD) {
      return onStay(car);
    }

    return onMove(car);
  }
```

sanghyuk-2i committed:
refactor: 의미없는 상수값 제거

yujo11 commented on __tests__/utils/isLapValid.test.js:
👍 👍 👍

yujo11 reviewed (approved):
안녕하세요 상혁님 ~

코멘트 남긴 내용에 더해 추가적인 개선까지 진행해주신 부분 아주 멋지네요!! 👍 👍 👍 

`determineCarMovement`도 제가 생각했던 것과 같은 방향으로 수정이 진행된거 같아요 ㅎㅎ 

자동차 미션은 여기서 이만 마쳐도 될거 같습니다! 첫 미션 멋지게 완료하신걸 축하드리고 끝까지 완주하셨으면 좋겠네요 !! 고생 많으셨습니다 상혁님 ~~~ 👍 👍 👍
